src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,323 @@
+                                         const std::vector<bool>& neighbors,
+                                         const Mn::Color3& color) {
+  // Using the data of a cubeSolid to create the voxel cube
+  assert(neighbors.size() >= 6);
+  Mn::Trade::MeshData cubeData = Mn::Primitives::cubeSolid();
+
+  // add cube to mesh
+  // midpoint of a voxel
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+
+  unsigned int sz = positions.size();
+  auto cubePositions = cubeData.positions3DAsArray();
+  auto cubeNormals = cubeData.normalsAsArray();
+  auto cubeIndices = cubeData.indices();
+
+  // TODO: Only add neccessary faces (based on neighborhood)
+  for (int i = 0; i < 24; i++) {
+    Mn::Vector3 vertOffset = cubePositions[i] * m_voxelSize / 2;
+    positions.push_back(vertOffset + mid);
+    // Set the normals to be weighted such that cubes look slightly curved
+    normals.push_back(cubePositions[i].normalized() * 1 / 4 +
+                      cubeNormals[i].normalized() * 3 / 4);
+    colors.push_back(color);
+  }
+
+  for (int i = 0; i < 6; i++) {
+    if (!neighbors[i]) {
+      for (int j = 0; j < 6; j++) {
+        indices.push_back(sz + cubeIndices[i * 6 + j][0]);
+      }
+    }
+  }
+}
+
+void VoxelGrid::addVectorToMeshPrimitives(std::vector<Mn::Vector3>& positions,
+                                          std::vector<Mn::Vector3>& normals,
+                                          std::vector<Mn::Color3>& colors,
+                                          std::vector<Mn::UnsignedInt>& indices,
+                                          const Mn::Vector3i& local_coords,
+                                          const Mn::Vector3& vec) {
+  Mn::Trade::MeshData coneData = Mn::Primitives::coneSolid(1, 3, 1.0f);
+
+  // midpoint of a voxel
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+
+  // add cone to mesh (tip of arrow)
+  unsigned int sz = positions.size();
+  const auto&& conePositions = coneData.positions3DAsArray();
+  const auto&& coneNormals = coneData.normalsAsArray();
+  const auto&& coneIndices = coneData.indices();
+
+  // Get rotation quaternion
+  Mn::Rad angle{static_cast<float>(
+      acos(Mn::Math::dot(vec.normalized(), Mn::Vector3(0, 1, 0))))};
+  Mn::Vector3 crossProduct = Mn::Math::cross(vec, Mn::Vector3(0, 1, 0));
+  Mn::Quaternion vecRotation{Mn::Math::IdentityInit};
+
+  if (vec[0] == 0 && vec[2] == 0) {
+    crossProduct = Mn::Vector3(0, 1, 0);
+  }
+  vecRotation = vecRotation.rotation(-angle, crossProduct.normalized());
+  for (const auto& ind : conePositions) {
+    positions.push_back(
+        vecRotation.transformVector(ind * Mn::Vector3(0.02, 0.04, 0.02) +
+                                    Mn::Vector3(0, 0.025, 0)) +
+        mid);
+    colors.push_back(Mn::Color3(0, .3, 1));
+  }
+
+  for (const auto& norm : coneNormals) {
+    normals.push_back(norm);
+  }
+
+  for (const auto&& index : coneIndices) {
+    indices.push_back(sz + index[0]);
+  }
```

Comment:
Step 4b -- and assuming you change the function to take `Cr::Containers::Array<Mn::UnsignedInt>& indices` instead of the vector, this can be 

```cpp
  for (const UnsignedInt index : coneIndices) {
    arrayAppend(indices, sz + index);
  }
```